getgauge/parser_parso.py

```diff
@@ -0,0 +1,140 @@
+            if content is None:
+                with open(file_path) as f:
+                    content = f.read()
+            py_tree = _parser.parse(content, path=file_path, error_recovery=False)
+            return ParsoPythonFile(file_path, py_tree)
+        except parso.parser.ParserSyntaxError as ex:
+            logging.error("Failed to parse %s:%d '%s'", file_path, ex.error_leaf.line, ex.error_leaf.get_code())
+
+    def __init__(self, file_path, py_tree):
+        self.file_path = file_path
+        self.py_tree = py_tree
+
+    def _span_from_pos(self, start_pos, end_pos):
+        return {
+            'start': start_pos[0],
+            'startChar': start_pos[1],
+            'end': end_pos[0],
+            'endChar': end_pos[1],
+        }
+
+    def _iter_step_func_decorators(self):
+        '''Find top level functions with step decorator in parsed file'''
+        for func in self.py_tree.iter_funcdefs():
+            for decorator in func.get_decorators():
+                if decorator.children[1].value == 'step':
+                    yield func, decorator
+                    break
+
+    def _step_decorator_args(self, decorator):
+        '''Get the arguments passed to step decorators converted to python objects'''
+        args = decorator.children[3:-2]
+        step = None
+        if len(args) == 1:
+            try:
+                step = ast.literal_eval(args[0].get_code())
+            except (ValueError, SyntaxError):
+                pass
+            if isinstance(step, six.string_types+(list,)):
+                return step
+            logging.error("Decorator step accepts either a string or a list of strings - %s:%d",
+                          self.file_path, decorator.start_pos[0])
+        else:
+            logging.error("Decorator step accepts only one argument - %s:%d",
+                          self.file_path, decorator.start_pos[0])
+
+    def iter_steps(self):
+        '''Iterate over steps in the parsed file'''
+        for func, decorator in self._iter_step_func_decorators():
+            step = self._step_decorator_args(decorator)
+            if step:
+                span = self._span_from_pos(decorator.start_pos, func.end_pos)
+                yield step, func.name.value, span
+
+    def _find_step_node(self, step_text):
+        '''Find the ast node which contains the text'''
+        for func, decorator in self._iter_step_func_decorators():
+            step = self._step_decorator_args(decorator)
+            arg_node = decorator.children[3]
+            if step == step_text:
+                return arg_node, func
+            elif isinstance(step, list) and step_text in step:
+                idx = step.index(step_text)
+                step_node = arg_node.children[1].children[idx * 2]
+                return step_node, func
+        return None, None
+
+    def refactor_step(self, old_text, new_text, move_param_from_idx):
```

Comment:
I think this method can be split into some small methods. It will make the code more readable and easy to understand.